File: src/wasmMain/resources/jsinterop/Document.js
konan.libraries.push({
    kvdom_Global_getDocument: function (resultArena) {
        return toArena(resultArena, document);
    },
    kvdom_Document_getElementById: function (arena, obj, idPtr, idLen, resultArena) {
        var id = toUTF16String(idPtr, idLen);
        var result = kotlinObject(arena, obj).getElementById(id);
        return toArena(resultArena, result);
    },
    kvdom_Document_createElement: function (arena, obj, tagPtr, tagLen, resultArena) {
        var tag = toUTF16String(tagPtr, tagLen);
        var result = kotlinObject(arena, obj).createElement(tag);
        return toArena(resultArena, result);
    },
    kvdom_Document_createTextNode: function (arena, obj, textPtr, textLen, resultArena) {
        var text = toUTF16String(textPtr, textLen);
        var result = kotlinObject(arena, obj).createTextNode(text);
        return toArena(resultArena, result);
    }
});

File: src/wasmMain/kotlin/lt/petuska/kvdom/jsexternal/Document.kt
package lt.petuska.kvdom.jsexternal

import kotlinx.wasm.jsinterop.*


open class JsDocument(arena: Arena, index: Object) : JsValue(arena, index), Document {
    override fun getElementById(elementID: String): DNode? {
        return allocateArena().let { newArena ->
            document_getElementById(
                arena, index,
                stringPointer(elementID),
                stringLengthBytes(elementID),
                newArena
            ).let {
                DNode(newArena, it)
            }
        }
    }

    override fun createElement(tag: String): DNode {
        return allocateArena().let { newArena ->
            document_createElement(
                arena, index,
                stringPointer(tag),
                stringLengthBytes(tag),
                newArena
            ).let {
                DNode(newArena, it)
            }
        }
    }

    override fun createTextNode(text: String): DNode {
        return allocateArena().let { newArena ->
            document_createTextNode(
                arena, index,
                stringPointer(text),
                stringLengthBytes(text),
                newArena
            ).let {
                DNode(newArena, it)
            }
        }
    }
}

actual val document = allocateArena().run {
    JsDocument(this, jsGetDocument(this)) as Document
}

@SymbolName("kvdom_Global_getDocument")
external fun jsGetDocument(resultArena: Arena): Object

@SymbolName("kvdom_Document_getElementById")
external fun document_getElementById(
    arena: Arena,
    index: Object,
    idPtr: Pointer,
    idLen: Int,
    resultArena: Arena
): Object

@SymbolName("kvdom_Document_createElement")
external fun document_createElement(
    arena: Arena,
    index: Object,
    tagPtr: Pointer,
    tagLen: Int,
    resultArena: Arena
): Object

@SymbolName("kvdom_Document_createTextNode")
external fun document_createTextNode(
    arena: Arena,
    index: Object,
    textPtr: Pointer,
    textLen: Int,
    resultArena: Arena
): Object